aten/src/ATen/TensorIterator.cpp

```diff
@@ -1420,6 +1429,13 @@ void TensorIteratorBase::build(TensorIteratorConfig& config) {
   mark_resize_outputs(config);
   // compute the result dtype and device
   compute_types(config);
+
+  if (config.coerce_to_scalar_) {
+    TORCH_INTERNAL_ASSERT(num_outputs_ == 1, "coerce_to_scalar is not supported for multi-output operators");
+    set_output(0, {}, {}, original_options(operands_[0]), names_);
+    return;
+  }
```

Comment:
@ngimel wdyt. This is pretty... hacky 😬. But it solves the problem of `smooth_l1_loss` unnecessarily allocating an intermediate tensor twice. And other ops that use TensorIterator to construct an intermediate that they then perform a reduction on will have this same problem when we try to port.

It still has the downside though of needing to run `TensorIterator` twice - once in the meta function and a second time in the impl function. Long term, maybe we can teach `TensorIterator` how to handle doing the computation on an intermediate tensor separately from the output, although that would probably be a more invasive change.